common/beerocks/bcl/include/bcl/beerocks_socket_event_loop.h

```diff
@@ -0,0 +1,127 @@
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _BEEROCKS_SOCKET_EVENT_LOOP_H_
+#define _BEEROCKS_SOCKET_EVENT_LOOP_H_
+
+#include "beerocks_event_loop.h"
+#include "network/socket.h"
+
+#include <unordered_map>
+
+namespace beerocks {
+
+/**
+ * @brief ePoll based implementation of the EventLoop interface.
+ * @see EventLoop
+ * 
+ * This class uses the Linux epoll APIs for monitoring the provided sockets for I/O operations.
+ * Timeout operations are achieved by using the timerfd mechanism, which delivers timer 
+ * expiration notifications via a file descriptor.
+ */
+class SocketEventLoop : public EventLoop {
+public:
+    /**
+     * @brief Class constructor.
+     * 
+     * Initializes an epoll file descriptor.
+     * 
+     * @param [in] timeout Sets the master timeout (in milliseconds) for the event loop.
+     */
+    explicit SocketEventLoop(std::chrono::milliseconds timeout = std::chrono::milliseconds::min());
+
+    /**
+     * @brief Class destructor.
+     */
+    virtual ~SocketEventLoop();
+
+    /**
+     * @see EventPoll::add_socket
+     */
+    virtual bool
+    add_socket(std::shared_ptr<Socket> socket, const EventLoop::EventHandlers &handlers,
+               std::chrono::milliseconds timeout = std::chrono::milliseconds::min()) override;
+
+    /**
+     * @see EventPoll::del_socket
+     */
+    virtual bool del_socket(std::shared_ptr<Socket> socket) override;
+
+    /**
+     * @see EventPoll::run
+     */
+    virtual int run() override;
+
+    /**
+     * @see EventPoll::die
+     */
+    virtual void die() override;
+
+protected:
+    /**
+     * @brief Main event loop method.
+     * 
+     * Executes the epoll_wait() function and processes ocurred events.
+     * 
+     * @return -1 on critical errors
+     * @return 0 on timeout without any socket events
+     * @return >0 number of socket events processed during the class to this method.
+     */
+    virtual int event_loop();
+
+private:
+    /**
+     * epoll file descriptor.
+     */
+    int m_epoll_fd = -1;
+
+    /**
+     * Event loop master timeout (used for the epoll_wait function).
+     */
+    std::chrono::milliseconds m_timeout = std::chrono::milliseconds::min();
+
+    /**
+     * Flag to mark whether the event loop is running.
+     */
+    bool m_running = false;
+
+    /**
+     * @brief Data structure representing a socket added to the poll.
+     * This structure groups all the information required for processing socket events.
+     */
+    struct EventData {
+        /**
+         * Socket event handler functions structure.
+         */
+        EventLoop::EventHandlers handlers;
+
+        /**
+         * Shared pointer to the socket object.
+         */
+        std::shared_ptr<Socket> socket = nullptr;
+
+        /**
+         * Shared pointer to the timeout file descriptor (wrapper in a Socket object).
+         */
+        std::shared_ptr<Socket> timeout = nullptr;
```

Comment:
> I think the template is the way to go here.

Even if we use templates, this is an abomination.

Reply:
That's just because `Socket` is the base FD wrapper we currently have.
After the `Socket` refactoring, perhaps we'll be able to change this to use a proper base-class.

Reply:
Looking at this again, there's no need for a wrapper class in this case.
Changed to `int`.